File: backend/app/agents/loop.py
"""Boucle agentique — gère le cycle tool_call → execution → réponse de l'agent."""
from __future__ import annotations

import json
import logging
from typing import AsyncGenerator

from app.agents.tools import TOOL_SCHEMAS, execute_tool
from app.config import settings
from app.services.ollama import OllamaClient

logger = logging.getLogger(__name__)

MAX_ITERATIONS = 20  # Sécurité : évite les boucles infinies


async def agentic_loop(
    model: str,
    messages: list[dict],
    system: str | None,
    enabled_tools: list[str] | None,
    ollama_base_url: str | None = None,
    ollama_api_mode: str | None = None,
    temperature: float = 0.7,
    max_tokens: int = 4096,
) -> AsyncGenerator[dict, None]:
    """Boucle agentique complète avec streaming des événements.

    Yields des dicts typés :
      - {"type": "text", "content": str}            → token de texte
      - {"type": "tool_start", "name": str, "args": dict}
      - {"type": "tool_result", "name": str, "result": dict}
      - {"type": "error", "message": str}
      - {"type": "done", "iterations": int}
    """
    client = OllamaClient(base_url=ollama_base_url, api_mode=ollama_api_mode)

    # Filtre les schémas selon les outils activés
    tools = (
        [t for t in TOOL_SCHEMAS if t["function"]["name"] in enabled_tools]
        if enabled_tools is not None
        else TOOL_SCHEMAS
    )

    history = list(messages)
    iterations = 0

    while iterations < MAX_ITERATIONS:
        iterations += 1
        logger.debug("Itération agent #%d", iterations)

        # Accumule le contenu texte et les tool calls pendant le stream
        text_buffer = ""
        tool_calls_buffer: list[dict] = []
        current_tc_idx = -1

        async for raw_chunk in client.stream_chat(
            model=model,
            messages=history,
            tools=tools if tools else None,
            system=system,
            temperature=temperature,
            max_tokens=max_tokens,
        ):
            try:
                chunk = json.loads(raw_chunk)
            except (json.JSONDecodeError, ValueError):
                continue

            for choice in chunk.get("choices", []):
                delta = choice.get("delta", {})
                finish = choice.get("finish_reason")

                # Texte courant
                if text := delta.get("content"):
                    text_buffer += text
                    yield {"type": "text", "content": text}

                # Tool call en cours de streaming
                if tcs := delta.get("tool_calls"):
                    for tc in tcs:
                        idx = tc.get("index", current_tc_idx + 1)
                        current_tc_idx = idx
                        while len(tool_calls_buffer) <= idx:
                            tool_calls_buffer.append({
                                "id": "", "type": "function",
                                "function": {"name": "", "arguments": ""}
                            })
                        entry = tool_calls_buffer[idx]
                        fn = tc.get("function", {})
                        entry["id"] = entry["id"] or tc.get("id", "")
                        entry["function"]["name"] = entry["function"]["name"] or fn.get("name", "")
                        entry["function"]["arguments"] += fn.get("arguments", "")

        # ── L'assistant a fini de streamer ──────────────────────────
        if not tool_calls_buffer:
            # Réponse texte pure → fin de la boucle
            if text_buffer:
                history.append({"role": "assistant", "content": text_buffer})
            break

        # ── Il y a des tool calls à exécuter ────────────────────────
        assistant_msg: dict = {
            "role": "assistant",
            "content": text_buffer or "",
            "tool_calls": [
                {
                    "id": tc["id"] or f"call_{i}",
                    "type": "function",
                    "function": {
                        "name": tc["function"]["name"],
                        "arguments": tc["function"]["arguments"],
                    },
                }
                for i, tc in enumerate(tool_calls_buffer)
                if tc["function"]["name"]
            ],
        }
        history.append(assistant_msg)

        # Exécute chaque outil
        tool_results_msgs: list[dict] = []
        for tc in assistant_msg["tool_calls"]:
            name = tc["function"]["name"]
            try:
                args = json.loads(tc["function"]["arguments"] or "{}")
            except (json.JSONDecodeError, ValueError):
                args = {}

            yield {"type": "tool_start", "name": name, "args": args, "call_id": tc["id"]}

            result = await execute_tool(name, args)

            yield {"type": "tool_result", "name": name, "result": result, "call_id": tc["id"]}

            tool_results_msgs.append({
                "role": "tool",
                "tool_call_id": tc["id"],
                "name": name,
                "content": json.dumps(result, ensure_ascii=False),
            })

        history.extend(tool_results_msgs)

    else:
        yield {"type": "error", "message": f"Limite d'itérations atteinte ({MAX_ITERATIONS})"}

    yield {"type": "done", "iterations": iterations}

File: backend-rs/src/agents/loop.rs
//! Agentic loop — streams tool_call -> execution -> response cycles as SSE events.

use futures::stream::Stream;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use sqlx::SqlitePool;
use std::pin::Pin;

use crate::config::Config;
use crate::services::mcp_proxy::{self, McpTool};
use crate::services::ollama::{get_setting_i64, OllamaClient};

use super::tools::{execute_tool, tool_schemas};

const MAX_ITERATIONS: usize = 20;

/// A typed event emitted by the agentic loop.
/// The chat route converts these into axum SSE Events.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum AgentEvent {
    #[serde(rename = "text")]
    Text { content: String },

    #[serde(rename = "tool_start")]
    ToolStart {
        name: String,
        args: Value,
        call_id: String,
    },

    #[serde(rename = "tool_result")]
    ToolResult {
        name: String,
        result: Value,
        call_id: String,
    },

    #[serde(rename = "error")]
    Error { message: String },

    #[serde(rename = "done")]
    Done { iterations: usize },
}

/// Run the agentic loop, yielding `AgentEvent` items.
///
/// The caller is responsible for converting these to SSE Events and
/// tracking state for DB persistence.
pub fn agentic_loop(
    client: OllamaClient,
    model: String,
    messages: Vec<Value>,
    system: Option<String>,
    enabled_tools: Option<Vec<String>>,
    temperature: f64,
    max_tokens: i64,
    config: Config,
    pool: SqlitePool,
) -> Pin<Box<dyn Stream<Item = AgentEvent> + Send>> {
    let stream = async_stream::stream! {
        // ── 1. Charge les outils internes ──
        let all_schemas = tool_schemas();
        let mut tools: Vec<Value> = match &enabled_tools {
            Some(names) => all_schemas
                .into_iter()
                .filter(|t| {
                    t.get("function")
                        .and_then(|f| f.get("name"))
                        .and_then(|n| n.as_str())
                        .map(|n| names.contains(&n.to_string()))
                        .unwrap_or(false)
                })
                .collect(),
            None => all_schemas,
        };

        // ── 2. Charge les outils MCP des serveurs activés ──
        let (mcp_schemas, mcp_registry) = mcp_proxy::load_mcp_tools(&config).await;
        if !mcp_schemas.is_empty() {
            tracing::info!("{} outils MCP injectés dans la boucle agentique", mcp_schemas.len());
            tools.extend(mcp_schemas);
        }

        let timeout_secs = get_setting_i64(&pool, "ollama_timeout", 300).await as u64;

        let mut history = messages;
        let mut iterations: usize = 0;

        loop {
            if iterations >= MAX_ITERATIONS {
                yield AgentEvent::Error {
                    message: format!("Iteration limit reached ({MAX_ITERATIONS})"),
                };
                break;
            }
            iterations += 1;
            tracing::debug!("Agent iteration #{iterations}");

            let tools_ref = if tools.is_empty() { None } else { Some(tools.as_slice()) };

            // Accumulate text and tool calls from the streaming response.
            let mut text_buffer = String::new();
            let mut tc_buffer: Vec<ToolCallAccum> = Vec::new();

            {
                use futures::StreamExt;
                let mut chunk_stream = client.stream_chat(
                    &model,
                    &history,
                    tools_ref,
                    system.as_deref(),
                    temperature,
                    max_tokens,
                    timeout_secs,
                );

                while let Some(chunk_result) = chunk_stream.next().await {
                    let raw = match chunk_result {
                        Ok(r) => r,
                        Err(e) => {
                            yield AgentEvent::Error { message: e };
                            break;
                        }
                    };

                    let chunk: Value = match serde_json::from_str(&raw) {
                        Ok(v) => v,
                        Err(_) => continue,
                    };

                    if let Some(choices) = chunk.get("choices").and_then(|c| c.as_array()) {
                        for choice in choices {
                            let delta = match choice.get("delta") {
                                Some(d) => d,
                                None => continue,
                            };

                            // Text content.
                            if let Some(text) = delta.get("content").and_then(|c| c.as_str()) {
                                if !text.is_empty() {
                                    text_buffer.push_str(text);
                                    yield AgentEvent::Text {
                                        content: text.to_string(),
                                    };
                                }
                            }

                            // Tool calls (streamed incrementally).
                            if let Some(tcs) = delta.get("tool_calls").and_then(|t| t.as_array()) {
                                for tc in tcs {
                                    let idx = tc
                                        .get("index")
                                        .and_then(|i| i.as_u64())
                                        .unwrap_or(tc_buffer.len() as u64)
                                        as usize;

                                    while tc_buffer.len() <= idx {
                                        tc_buffer.push(ToolCallAccum::default());
                                    }
                                    let entry = &mut tc_buffer[idx];

                                    if let Some(id) = tc.get("id").and_then(|v| v.as_str()) {
                                        if entry.id.is_empty() {
                                            entry.id = id.to_string();
                                        }
                                    }
                                    if let Some(func) = tc.get("function") {
                                        if let Some(name) =
                                            func.get("name").and_then(|n| n.as_str())
                                        {
                                            if entry.name.is_empty() {
                                                entry.name = name.to_string();
                                            }
                                        }
                                        if let Some(args_frag) =
                                            func.get("arguments").and_then(|a| a.as_str())
                                        {
                                            entry.arguments.push_str(args_frag);
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }

            // -- The assistant finished streaming for this iteration --

            if tc_buffer.is_empty() {
                // Pure text response — done.
                if !text_buffer.is_empty() {
                    history.push(json!({"role": "assistant", "content": text_buffer}));
                }
                break;
            }

            // Build finalized tool calls with stable IDs.
            let finalized: Vec<(String, String, String)> = tc_buffer
                .iter()
                .enumerate()
                .filter(|(_, tc)| !tc.name.is_empty())
                .map(|(i, tc)| {
                    let id = if tc.id.is_empty() {
                        format!("call_{i}")
                    } else {
                        tc.id.clone()
                    };
                    (id, tc.name.clone(), tc.arguments.clone())
                })
                .collect();

            // Append assistant message (with tool_calls) to history.
            let tc_values: Vec<Value> = finalized
                .iter()
                .map(|(id, name, arguments)| {
                    json!({
                        "id": id,
                        "type": "function",
                        "function": {
                            "name": name,
                            "arguments": arguments,
                        }
                    })
                })
                .collect();

            history.push(json!({
                "role": "assistant",
                "content": text_buffer,
                "tool_calls": tc_values,
            }));

            // Execute each tool and feed results back.
            for (call_id, name, raw_args) in &finalized {
                let args: Value = serde_json::from_str(raw_args).unwrap_or(json!({}));

                yield AgentEvent::ToolStart {
                    name: name.clone(),
                    args: args.clone(),
                    call_id: call_id.clone(),
                };

                // Route : outil MCP ou outil interne ?
                let result = if name.starts_with("mcp__") {
                    // Trouve l'outil MCP dans le registre
                    if let Some(mcp_tool) = mcp_registry.iter().find(|t| t.qualified_name == *name) {
                        mcp_proxy::execute_mcp_tool(mcp_tool, &args).await
                    } else {
                        json!({"error": format!("MCP tool not found: {name}")})
                    }
                } else {
                    execute_tool(name, args, &config, &pool).await
                };

                yield AgentEvent::ToolResult {
                    name: name.clone(),
                    result: result.clone(),
                    call_id: call_id.clone(),
                };

                history.push(json!({
                    "role": "tool",
                    "tool_call_id": call_id,
                    "name": name,
                    "content": serde_json::to_string(&result).unwrap_or_default(),
                }));
            }
        }

        yield AgentEvent::Done { iterations };
    };

    Box::pin(stream)
}

// ------------------------------------------------------------------
// Internal accumulator for streamed tool calls
// ------------------------------------------------------------------

#[derive(Default, Clone)]
struct ToolCallAccum {
    id: String,
    name: String,
    arguments: String,
}
